### DISCUSSION TITLE
Behaviour of joins with TypeDecorators

### DISCUSSION THREAD

**Discussion by reritom** [sqlalchemy/sqlalchemy#8970] [Usage Questions] [answered]
When performing a query which has a join on another model, where that other model has a TypeDecorator field, the `process_result_value` gets called even when there are no instances of that model.

```py
from sqlalchemy import create_engine, MetaData, TypeDecorator, Column, ForeignKey, Integer
from sqlalchemy.types  import JSON
from sqlalchemy.orm import Session, declarative_base,relationship
 
engine = create_engine("sqlite+pysqlite:///:memory:",future=True)
Meta = MetaData(bind=engine)
session = Session(engine)
Base = declarative_base(metadata=Meta)


class Custom(TypeDecorator):
    impl = JSON

    cache_ok = True

    def process_bind_param(self, value, dialect):
        return {"wrapped": value}

    def process_result_value(self, value, dialect):
        return value["wrapped"]


class Child(Base):
    __tablename__ = "child"
    id = Column(Integer, primary_key=True)
    custom = Column(Custom, nullable=False)
    associations = relationship("Association", back_populates="child")

class Association(Base):
    __tablename__ = "association"
    id = Column(Integer, primary_key=True)

    parent_id = Column(ForeignKey("parent.id"), nullable=False)
    parent = relationship("Parent", foreign_keys=parent_id, back_populates="associations")

    child_id = Column(ForeignKey("child.id"), nullable=False)
    child = relationship("Child", foreign_keys=child_id, back_populates="associations", lazy="joined")


class Parent(Base):
    __tablename__ = "parent"
    id = Column(Integer, primary_key=True)
    associations = relationship("Association", back_populates="parent", lazy="joined")


Meta.create_all()

if True: # Toggle to False for the other case
    # This case works (where there is a row for the other model
    parent = Parent()
    child = Child()
    association = Association(parent=parent, child=child)
    session.add(association)
    print(session.query(Parent).first().associations[0].child)
else:
    # This doesnt work. There are no Child rows defined, but the `process_result_value` on the Custom field on the child gets called.
    # If we remove the `lazy="joined"` from the `Parent.associations` relationship, there is no error.
    parent = Parent()
    session.add(parent)
    print(session.query(Parent).first())
```

The error is:
```
Traceback (most recent call last):
  File "/Users/me/Documents/Projects/int/main.py", line 57, in <module>
    print(session.query(Parent).first())
  File "/Users/me/Documents/Projects/int/.venv/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 2824, in first
    return self.limit(1)._iter().first()
  File "/Users/me/Documents/Projects/int/.venv/lib/python3.9/site-packages/sqlalchemy/engine/result.py", line 1498, in first
    return self._only_one_row(
  File "/Users/me/Documents/Projects/int/.venv/lib/python3.9/site-packages/sqlalchemy/engine/result.py", line 559, in _only_one_row
    row = onerow(hard_close=True)
  File "/Users/me/Documents/Projects/int/.venv/lib/python3.9/site-packages/sqlalchemy/engine/result.py", line 1386, in _fetchone_impl
    return self._real_result._fetchone_impl(hard_close=hard_close)
  File "/Users/me/Documents/Projects/int/.venv/lib/python3.9/site-packages/sqlalchemy/engine/result.py", line 1801, in _fetchone_impl
    row = next(self.iterator, _NO_ROW)
  File "/Users/me/Documents/Projects/int/.venv/lib/python3.9/site-packages/sqlalchemy/orm/loading.py", line 147, in chunks
    fetch = cursor._raw_all_rows()
  File "/Users/me/Documents/Projects/int/.venv/lib/python3.9/site-packages/sqlalchemy/engine/result.py", line 393, in _raw_all_rows
    return [make_row(row) for row in rows]
  File "/Users/me/Documents/Projects/int/.venv/lib/python3.9/site-packages/sqlalchemy/engine/result.py", line 393, in <listcomp>
    return [make_row(row) for row in rows]
  File "/Users/me/Documents/Projects/int/.venv/lib/python3.9/site-packages/sqlalchemy/sql/type_api.py", line 1707, in process
    return process_value(impl_processor(value), dialect)
  File "/Users/me/Documents/Projects/int/main.py", line 20, in process_result_value
    return value["wrapped"]
TypeError: 'NoneType' object is not subscriptable
```
I would expect there to be no key error because this method shouldn't be getting called because there is no row.

Am I misunderstanding something?

Version: SQLAlchemy==1.4.45

**Comment by zzzeek** [comment, marked as answer]
If I understand correctly, there's a LEFT OUTER JOIN which means the related rows may have None for all columns, but nonetheless that is a row.  The type objects will be run for their columns within each row so this would be expected. 

in the examples at https://docs.sqlalchemy.org/en/14/core/custom_types.html#typedecorator-recipes note that all the TypeDecorator objects accommodate for None.

**Reply by reritom** [reply]
Thanks for the quick feedback. I'm not familiar with this. To clarify your response, you are saying that even though there are no rows for the model Child, the join from the Parent results in a row with None for each column? I think I am misunderstanding your response.

In the given example, the error case has only 1 row for the Parent model, and no rows for the other models. So it seems odd that it is trying to process the `Child.custom` when there are no Child rows

**Reply by zzzeek** [reply]
> Thanks for the quick feedback. I'm not familiar with this. To clarify your response, you are saying that even though there are no rows for the model Child, the join from the Parent results in a row with None for each column? I think I am misunderstanding your response.

you are using "lazy='joined'" for both links, which means that the Child object is loaded using LEFT OUTER JOIN:

    SELECT parent.*, association.*, child.* from parent LEFT OUTER JOIN association ON <onclause> LEFT OUTER JOIN child ON <onclause>


in SQL, SELECT always returns a fixed set of columns.   In LEFT OUTER JOIN, the columns for the "outer joined" table, if no row matches the parent row, are sent as NULL.

So when one row is returned for "parent", we get a row like:

    (1, NULL, NULL, NULL, NULL, NULL)

where parent.id == 1, assocation's three columns are NULL, child's three columns are NULL

the Core SELECT statement has processors set up as:

    Parent.id, Association.id, Association.parent_id, Association.child_id, Child.id, Child.custom

SQLAlchemy's row handling routines pass NULL to result processors, as there's lots of kinds of types that want to do something special for NULL.   In 1.4 these processors [run unconditionally](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#proxying-behavior-goes-away-was-also-unnecessary-in-modern-usage) even though the ORM here is likely not even looking at these columns.





> 
> In the given example, the error case has only 1 row for the Parent model, and no rows for the other models. So it seems odd that it is trying to process the `Child.custom` when there are no Child rows


your SELECT statement has columns for all three entities since you asked it to use "joined" eager loading.  use some other kind of eager loading and the NULL columns will no longer be fetched.

**Reply by reritom** [reply]
Thank you. That reply is concise, answers my questions, and taught me something.
